Utilities/bootstrap

```diff
@@ -167,14 +167,16 @@ def add_test_args(parser):
 
 def parse_global_args(args):
     """Parses and cleans arguments necessary for all actions."""
-    args.build_dir = os.path.abspath(args.build_dir)
-    args.project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
-    args.tsc_source_dir = os.path.join(args.project_root, "..", "swift-tools-support-core")
-    args.yams_source_dir = os.path.join(args.project_root, "..", "yams")
-    args.swift_argument_parser_source_dir = os.path.join(args.project_root, "..", "swift-argument-parser")
-    args.swift_driver_source_dir = os.path.join(args.project_root, "..", "swift-driver")
-    args.swift_crypto_source_dir = os.path.join(args.project_root, "..", "swift-crypto")
-    args.source_root = os.path.join(args.project_root, "Sources")
+    args.build_dir                            = os.path.abspath(args.build_dir)
+    args.build_dirs = {}
```

Comment:
The CI is failing because of this line, which resets `build_dirs` when `parse_test_args()` is called before running the tests. Are you sure setting it to an empty dictionary here is necessary?

Reply:
Thanks for figuring that out — perhaps both `build_dirs` and `source_dirs` (below) need to be initialized as dictionaries earlier, so that the code here just sets keys in them rather than resetting them.

Reply:
> The CI is failing because of this line, which resets `build_dirs` when `parse_test_args()` is called before running the tests.

Thanks for figure this out. 

> Are you sure setting it to an empty dictionary here is necessary?

I think initialising `build_dirs` and `source_dir` as empty dictionaries are necessary, because Python will raise an error if we try to insert a value into a dictionary before initialising it.

> perhaps both `build_dirs` and `source_dirs` (below) need to be initialized as dictionaries earlier, so that the code here just sets keys in them rather than resetting them.

I can't find a good spot to initialize them earlier, so instead, I changed the code to test if both dictionaries already exist, and only initialize them to empty dictionaries if they don't exist. (19e458386da0788b3146321b0ededb9b6363a455)

Reply:
I agree that initializing them is necessary, but I think it could be done unconditionally at the end of `main`, after `args = parser.parse_args()` but before `args.func(args)`.

Reply:
But it's also fine to have the initialized on-demand, even though it's a bit more text.